drakrun/drakrun/lib/postprocessing/capa_plugin/capa_processor.py

```diff
@@ -0,0 +1,428 @@
+    ttp_block.update({"occurrences": occurrences})
+
+    return ttp_block
+
+
+def construct_ttp_blocks(
+    rules: capa.rules.RuleSet,
+    capabilities_per_file: List[Tuple[Path, capa.engine.MatchResults, Any]],
+    filter_function=lambda rule: rule,
+) -> Iterator[Dict[str, Any]]:
+
+    # construct a ttp block for each extracted capability
+    for _, capabilities in capabilities_per_file:
+        for name, addresses in capabilities.items():
+            if filter_function(rules[name]):
+                yield construct_ttp_block(rules[name], addresses)
+
+
+def capa_analysis(analysis_dir: Path) -> None:
+    # check and prepare the rules folder
+    if not check_rules_directory_exist(capa_rules_dir):
+        # in case of a missing/empty rules folder, clone the official capa rules
+        logger.exception("capa rules directory is empty or non-existant")
+
+    # get rules and filter them
+    rules = get_rules([capa_rules_dir])
+    rules = filter_rules(
+        rules, filter_function=lambda rule: rule.meta.get("att&ck", None)
+    )  # select only rules with an att&ck entry specified
+
+    # get malware-related pids if requested by configuration
+    malware_pids = None
+    if analyze_malware_pids_only:
+        malware_pids = list(
+            get_malware_processes(
+                metadata_path=analysis_dir / "metadata.json",
+                inject_path=analysis_dir / "inject.log",
+                pstree_path=analysis_dir / "process_tree.json",
+            )
+        )
+
+    # make sure either static or dynamic capability extraction is on
+    assert perform_dynamic_analysis or perform_static_analysis
+
+    # extract capabilities from the Drakvuf report
+    if perform_dynamic_analysis:
+        dynamic_capabilities = dynamic_capa_analysis(
+            analysis_dir, rules, malware_pids=malware_pids
+        )
+
+        # write the extracted TTPs to the analysis dir
+        with (analysis_dir / "ttps.json").open("wb") as f:
+            for ttp in construct_ttp_blocks(
+                rules,
+                [dynamic_capabilities],
+                filter_function=lambda rule: rule.meta.get("att&ck", None),
+            ):
+                f.write(orjson.dumps(ttp))
+                f.write(b"\n")
+
+    # extract capabilities from the memory dumps
+    if perform_static_analysis:
+        static_capabilities_per_file = static_memory_dumps_capa_analysis(
+            analysis_dir, rules, malware_pids=malware_pids
+        )
+
+        # create a folder containing the TTPs corresponding to each dump
+        dumps_ttp_path = analysis_dir / "dumps_ttp"
+        dumps_ttp_path.mkdir(parents=True, exist_ok=True)
+
+        # dump the TTPs for each memdump into a jsonl file
+        for dump_name, static_capabilities in static_capabilities_per_file:
+            with (dumps_ttp_path / dump_name).open("wb") as f:
+                for ttp in construct_ttp_blocks(rules, [static_capabilities]):
```

Comment:
Types don't add up here. [static_capabilities] are of type List[MatchResults] (aka `List[Mapping[str, List[Tuple[Address, Result]]]])`, but the function expects `List[Tuple[Path, capa.engine.MatchResults, Any]]`

They are probably compatible, but signature should be fixed.

Reply:
This should be fixed now, I'm reconstructing the tuple now after I grab the dump's name.